File: src/Entity/Achievement.php
<?php

namespace App\Entity;

use App\Repository\AchievementRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: AchievementRepository::class)]
class Achievement
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $code = null;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $name = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $description = null;

    #[ORM\Column(nullable: true)]
    private ?array $rewards = null;

    #[ORM\Column(nullable: true)]
    private ?bool $isPremium = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getCode(): ?string
    {
        return $this->code;
    }

    public function setCode(?string $code): static
    {
        $this->code = $code;

        return $this;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(?string $name): static
    {
        $this->name = $name;

        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;

        return $this;
    }

    public function getRewards(): ?array
    {
        return $this->rewards;
    }

    public function setRewards(?array $rewards): static
    {
        $this->rewards = $rewards;

        return $this;
    }

    public function isPremium(): ?bool
    {
        return $this->isPremium;
    }

    public function setIsPremium(?bool $isPremium): static
    {
        $this->isPremium = $isPremium;

        return $this;
    }
}

File: flintf/lib/models/achievement.dart
class Achievement {
  final String code;
  final String name;
  final String description;
  final Map<String, dynamic> rewards;
  final bool isPremium;
  final bool isUnlocked;

  Achievement({
    required this.code,
    required this.name,
    required this.description,
    required this.rewards,
    required this.isPremium,
    this.isUnlocked = false,
  });

  factory Achievement.fromJson(Map<String, dynamic> json, {Set<String> unlocked = const {}}) {
    return Achievement(
      code: json['code'],
      name: json['name'],
      description: json['description'],
      rewards: Map<String, dynamic>.from(json['rewards']),
      isPremium: json['isPremium'],
      isUnlocked: unlocked.contains(json['code']),
    );
  }
}
